### ISSUE TITLE
AttributeError: 'Dataframe' object has no attribute '_data'; dat.save() destroys original save file

### ISSUE THREAD

**Issue by danielsvedberg** [nubs01/blechpy#31]
<pre><font color="#008700">In [</font><font color="#8AE234"><b>25</b></font><font color="#008700">]: </font>dat2.save()                                                                                      
<font color="#CC0000">---------------------------------------------------------------------------</font>
<font color="#CC0000">AttributeError</font>                            Traceback (most recent call last)
<font color="#4E9A06">&lt;ipython-input-25-5bdf4b602613&gt;</font> in <font color="#06989A">&lt;module&gt;</font>
<font color="#4E9A06">----&gt; 1</font><font color="#CC0000"> </font>dat2<font color="#3465A4">.</font>save<font color="#3465A4">()</font>

<font color="#4E9A06">~/Documents/CodeVault/blechpy/blechpy/datastructures/objects.py</font> in <font color="#06989A">save</font><font color="#3465A4">(self)</font>
<font color="#4E9A06"><b>     37</b></font>     <font color="#4E9A06">def</font> save<font color="#3465A4">(</font>self<font color="#3465A4">):</font>
<font color="#4E9A06"><b>     38</b></font>         <font color="#4E9A06">with</font> open<font color="#3465A4">(</font>self<font color="#3465A4">.</font>save_file<font color="#3465A4">,</font> <font color="#3465A4">&apos;wb&apos;)</font> <font color="#4E9A06">as</font> f<font color="#3465A4">:</font>
<font color="#4E9A06">---&gt; 39</font><font color="#CC0000">             </font>pickle<font color="#3465A4">.</font>dump<font color="#3465A4">(</font>self<font color="#3465A4">,</font> f<font color="#3465A4">)</font>
<font color="#4E9A06"><b>     40</b></font>             print<font color="#3465A4">(&apos;Saved %s to %s\n&apos;</font> <font color="#3465A4">%</font> <font color="#3465A4">(</font>self<font color="#3465A4">.</font>data_name<font color="#3465A4">,</font> self<font color="#3465A4">.</font>save_file<font color="#3465A4">))</font>
<font color="#4E9A06"><b>     41</b></font> 

<font color="#4E9A06">~/anaconda3/envs/blechpy/lib/python3.7/site-packages/pandas/core/generic.py</font> in <font color="#06989A">__getstate__</font><font color="#3465A4">(self)</font>
<font color="#4E9A06"><b>   1932</b></font>         meta <font color="#3465A4">=</font> <font color="#3465A4">{</font>k<font color="#3465A4">:</font> getattr<font color="#3465A4">(</font>self<font color="#3465A4">,</font> k<font color="#3465A4">,</font> <font color="#4E9A06">None</font><font color="#3465A4">)</font> <font color="#4E9A06">for</font> k <font color="#4E9A06">in</font> self<font color="#3465A4">.</font>_metadata<font color="#3465A4">}</font>
<font color="#4E9A06"><b>   1933</b></font>         return dict(
<font color="#4E9A06">-&gt; 1934</font><font color="#CC0000">             </font>_data<font color="#3465A4">=</font>self<font color="#3465A4">.</font>_data<font color="#3465A4">,</font>
<font color="#4E9A06"><b>   1935</b></font>             _typ<font color="#3465A4">=</font>self<font color="#3465A4">.</font>_typ<font color="#3465A4">,</font>
<font color="#4E9A06"><b>   1936</b></font>             _metadata<font color="#3465A4">=</font>self<font color="#3465A4">.</font>_metadata<font color="#3465A4">,</font>

<font color="#4E9A06">~/anaconda3/envs/blechpy/lib/python3.7/site-packages/pandas/core/generic.py</font> in <font color="#06989A">__getattr__</font><font color="#3465A4">(self, name)</font>
<font color="#4E9A06"><b>   5268</b></font>             <font color="#4E9A06">or</font> name <font color="#4E9A06">in</font> self<font color="#3465A4">.</font>_accessors
<font color="#4E9A06"><b>   5269</b></font>         ):
<font color="#4E9A06">-&gt; 5270</font><font color="#CC0000">             </font><font color="#4E9A06">return</font> object<font color="#3465A4">.</font>__getattribute__<font color="#3465A4">(</font>self<font color="#3465A4">,</font> name<font color="#3465A4">)</font>
<font color="#4E9A06"><b>   5271</b></font>         <font color="#4E9A06">else</font><font color="#3465A4">:</font>
<font color="#4E9A06"><b>   5272</b></font>             <font color="#4E9A06">if</font> self<font color="#3465A4">.</font>_info_axis<font color="#3465A4">.</font>_can_hold_identifiers_and_holds_name<font color="#3465A4">(</font>name<font color="#3465A4">):</font>

<font color="#CC0000">AttributeError</font>: &apos;DataFrame&apos; object has no attribute &apos;_data&apos;
</pre>

This has been a long running issue that I think I've finally pinned down. 
The trigger is always moving a file from one machine to another and then attempting to _change_root() and then save. Sometimes this can cause problems like HMM runtime errors as well.

In general if you encounter this problem make sure you BACK UP YOUR ORIGINAL .P FILE AND DON'T FUCK WITH IT because it is a HUGE pain in the ass to try and reconstitute the dataset after it's gone for good. This problem doesn't necessarily mean your .p file is corrupted so HOLD ON TO IT. 

[This stackexchange thread ](https://stackoverflow.com/questions/63559007/unpickling-dictionary-that-holds-pandas-dataframes-throws-attributeerror-dataf) suggests the cause is pandas version incompatibility, and recommends upgrading to pandas version >=1.1.0. **Upgrading my pandas version from 1.0.1 -> 1.1.0 has helped me rescue an older dataset suffering from the problem. Will see if this affects newer datasets.**

**Comment by danielsvedberg** [comment]
Just to leave a record of my progress/notes on this issue since it's complicated and I'll want to look back in the future:

I was trying to run HMMs on some old datasets and was confronted with a runtime error where the HMMs wouldn't display progress messages and the CPUs would demonstrate zero activity. Running HMMs with parallel = False would throw the error "index error: empty selections not allowed", seemingly related to phmm.get_spike_data()/h5io.get_spike_data(), and perhaps phmm.query_units(). Initially I traced this down to the fact that the .p file was getting wiped when the root would change, which is why I troubleshot the pandas-version incompatibility issues. 

After successfully using pandas 1.1.0 to read in the .p files, I was still getting the error. I identified that under pandas 1.1.0, dat.get_unit_table() returns the unit table fully-relabeled with zeros and None in various columns. This is purely a read-error, upgrading Pandas to 1.3.5 solved this issue and seems to not cause other problems...let's wait and see if this is the case. With these datasets I am at least able to _change_root(), save(), get_unit_table, and get electrode_mapping without issues. 

I also had to refactor my 'area' column since in the past I was labeling my CAR groups as 'GC_R' and 'GC_L' and recently changed to 'GC' for both sides, which is a 'me' problem that may have been driving some of these issues. 

I was also upgraded joblib from 0.14.1 to 0.17.0 and psutil from 5.7.0 to 5.9.0 due to an unrelated problem where blech_clust_run() would hang because of memory management issues, and this seems  to not be causing problems with running the HMMs. 

Currently running the HMMs on jetstream on a "problem" dataset, although it seems I have more troubleshooting to do with the other one. I also need to see if this Pandas upgrade will cause me problems with the newer datasets, which is why I'm taking meticulous notes here so I can backtrack if need be. 

Pending successful testing, I'll pull the new Pandas and joblib requirements into the next blechpy version.
